Tf widget

StasJ opened the pull request.

StasJ commented on share/palettes/CoolWarm.tf3:
```diff
@@ -1,31 +1,16 @@
 <OpacityScale Type="Double">
   1 
 </OpacityScale>
-<ColorMapParams>
```
I did not change the colormap file format. None of the current divergent colormaps specified whether they used whitepoints or not which they should. This was being implicitly set to use whitepoint in the getter function. Since there is no way to set a default value in the params database without writing a getter/setter I made sure the colormaps specified this. I also fixed some of the default control points which were at odd intervals.

StasJ replied:
One thing to clarify is that the _init() functions allow you to specify initial values when creating a new param but not default values.

clyne replied:
Sounds good.

StasJ commented on apps/vaporgui/Histo.cpp:
```diff
@@ -17,140 +17,414 @@
+    _range = _maxMapData - _minMapData;
 
-    _binArray = new long[_numBins];
+    _binArray = new unsigned int[_numBins];
```
Histo does not need a copy constructor. It is only there for legacy purposes and we can remove it once the old TFWidget is removed. I can still convert it to a std::vector if you prefer.

clyne replied:
No need to change it if it is being removed.

StasJ commented on apps/vaporgui/Histo.cpp:
```diff
@@ -17,140 +17,414 @@
+    } else if (val > _maxMapData) {
+        _numSamplesAbove++;
+        if (_above) {
+            int index = (val-_maxMapData)/(_maxData-_maxMapData) * _nBinsAbove;
```
If there is a div by zero here then the object is invalid so it does not matter. You can make this a div by zero in the GUI and everything works fine.

clyne replied:
Then add an assert, or clamp index to a valid range if possible.

StasJ replied:
It is not possible to clamp to a valid range because if there were a divide by zero then there would be no valid range. I added an assert.

StasJ commented on apps/vaporgui/Histo.cpp:
```diff
@@ -17,140 +17,414 @@
+    
+    double dx = (maxExts[X]-minExts[X]) / SAMPLE_RATE;
+    double dy = (maxExts[Y]-minExts[Y]) / SAMPLE_RATE;
+    double dz = (maxExts[Z]-minExts[Z]) / SAMPLE_RATE;
```
I did not write this code, I just moved it from the old TFWidget. The old code was written in such a way that this is only called for 3D variables

clyne replied:
About adding an an assert:

VAssert(minExts.size() >= 3 && maxExts.size() >= 3)

StasJ replied:
done

StasJ commented on apps/vaporgui/ParamsMenuItems.h:
```diff
@@ -0,0 +1,70 @@
+    void Update(VAPoR::ParamsBase *p);
+    
+private:
+    int getValueForIndex(int index) const;
```
done

StasJ commented on apps/vaporgui/ParamsWidgets.cpp:
```diff
@@ -0,0 +1,311 @@
+    _params = p;
+    
+    QColor color = VectorToQColor(p->GetValueDoubleVec(_tag));
+//    QString style = "background-color: " + color.name() + "; ";
```
done

StasJ commented on apps/vaporgui/ParamsWidgets.cpp:
```diff
@@ -0,0 +1,311 @@
+}
+
+void ParamsWidgetRange::valueChangedSlot()
+{
```
done

StasJ commented on apps/vaporgui/ParamsWidgets.cpp:
```diff
@@ -0,0 +1,311 @@
+    _color->setColor(color);
+}
+
+//void ParamsWidgetColor::pressed()
```
done

StasJ commented on apps/vaporgui/TFHistogramWidget.cpp:
```diff
@@ -0,0 +1,168 @@
+        switch (scaling) {
+            default:
+            case ScalingType::Linear:
+                bin /= maxBin;
```
`Histo::getMaxBinSizeBetweenIndices` && `Histo::getMaxBinSize` guarantee to return a non-zero number although I missed a 1/log(1) so I fixed that and added a zero check too.

clyne replied:
You have code that checks for maxBin == to zero:

    const float logMaxBin = maxBin == 0 ? 1 : logf(maxBin);

So it looks like maxBin can indeed be zero.

StasJ replied:
Good catch but no, it can't. That code is supposed to be
```
const float logMaxBin = maxBin == 1 ? 1 : logf(maxBin);
```
because `log(1) = 0`

StasJ commented on apps/vaporgui/ParamsMenuItems.h:
```diff
@@ -0,0 +1,70 @@
+// ******************************
+
+
+class ParamsMenuItem : public QAction {
```
Done

StasJ commented on apps/vaporgui/ParamsWidgets.h:
```diff
@@ -0,0 +1,169 @@
+#include <vapor/ParamsBase.h>
+#include <cassert>
+
+class ParamsWidget : public QWidget {
```
Done

StasJ commented on apps/vaporgui/QColorWidget.h:
```diff
@@ -0,0 +1,27 @@
+
+class QStylePainter;
+
+class QColorWidget : public QLineEdit {
```
What sort of documentation should I put?

```
//! \class QColorWidget
//! A Qt widget that displays a color and allows changing said color.
```

I don't think it adds anything.

StasJ replied:
Done

StasJ commented on apps/vaporgui/QPaintUtils.h:
```diff
@@ -0,0 +1,14 @@
+class QPainter;
+class QRect;
+
+namespace QPaintUtils {
```
Same here. The documentation would be "A collection of utilities for QPainter"

clyne replied:
For starters, knowing that these functions operate on Qt's QPaint objects would be helpful. 

Knowing what DropShadow(), InnerShadow(), etc. do would also be helpful. 

One of the confusing aspects of the naming convention used by the Q* widgets is that it is hard to know if a Q* widget is a Qt widget, or a custom (VAPOR) widget. Hence, finding documentation is non-trivial....

StasJ replied:
It is neither a `QWidget` nor a "Vapor Widget". Qt follows the convention of naming its things Q... even if they're not QWidgets. We have a bunch of classes that are called Q... so my naming was not unprecedented but I can rename to VPaintUtils or something if you'd like. I added documentation on what these things do in the meantime.

StasJ commented on apps/vaporgui/QRangeSlider.h:
```diff
@@ -0,0 +1,47 @@
+
+class QStylePainter;
+
+class QRangeSlider : public QSlider {
```
```
//! \class QRangeSlider
//! A QSlider that selects a range.
```

StasJ replied:
Done

StasJ commented on apps/vaporgui/QRangeSliderTextCombo.cpp:
```diff
@@ -0,0 +1,87 @@
+#include <QDoubleValidator>
+#include <vapor/VAssert.h>
+
+QRangeSliderTextCombo::QRangeSliderTextCombo()
```
This is the implementation

StasJ commented on apps/vaporgui/QRangeSliderTextCombo.h:
```diff
@@ -0,0 +1,36 @@
+#include <QLineEdit>
+#include <QWidget>
+
+class QRangeSliderTextCombo : public QWidget {
```
Done

clyne replied:
What do the signals do? They're not conventional Qt signals.

StasJ replied:
Done

StasJ commented on apps/vaporgui/TFColorInfoWidget.h:
```diff
@@ -0,0 +1,33 @@
+    class RenderParams;
+}
+
+class TFColorInfoWidget : public TFInfoWidget {
```
Done

clyne replied:
I don't see any documentation

StasJ replied:
I added documentation to `TFInfoWidget`. `TFColorInfoWidget` is an internal component of the `TFEditor` and it is just a `TFInfoWidget` that is specialized for the `TFColorMap`

StasJ commented on apps/vaporgui/TFColorWidget.h:
```diff
@@ -0,0 +1,103 @@
+class ParamsDropdownMenuItem;
+class ParamsCheckboxMenuItem;
+
+class TFColorMap : public TFMap {
```
I added documentation for TFMap. Here the documentation would only state "A TFMap for the color parameter" which doesn't add anything.

StasJ commented on apps/vaporgui/TFHistogramWidget.cpp:
```diff
@@ -0,0 +1,168 @@
+    
+    p.fillRect(rect(), Qt::lightGray);
+    
+//    QMatrix m;
```
Done

StasJ commented on apps/vaporgui/TFHistogramWidget.h:
```diff
@@ -0,0 +1,57 @@
+
+class ParamsDropdownMenuItem;
+
+class TFHistogramMap : public TFMap {
```
I added documentation for TFMap. Here the documentation would only state "A TFMap showing the histogram" which doesn't add anything.

StasJ commented on apps/vaporgui/TFMapGroupWidget.h:
```diff
@@ -0,0 +1,59 @@
+
+
+
+class TFMapGroupWidget : public QWidget {
```
Done

StasJ commented on apps/vaporgui/TFOpacityWidget.cpp:
```diff
@@ -0,0 +1,365 @@
+static vec2 qvec2(const QPointF &qp) { return vec2(qp.x(), qp.y()); }
+static QPointF qvec2(const vec2 &v) { return QPointF(v.x, v.y); }
+
+vec2 Project(vec2 a, vec2 b, vec2 p)
```
Done

StasJ commented on apps/vaporgui/TFOpacityWidget.cpp:
```diff
@@ -0,0 +1,365 @@
+    return n * t + a;
+}
+
+float DistanceToLine(vec2 a, vec2 b, vec2 p)
```
Done

StasJ commented on apps/vaporgui/VLineItem.h:
```diff
@@ -0,0 +1,11 @@
+#include <QWidget>
+#include <string>
+
+class VLineItem : public QWidget {
```
Done

StasJ commented on apps/vaporgui/TFMapWidget.h:
```diff
@@ -0,0 +1,123 @@
+class TFInfoWidget;
+class TFMapWidget;
+
+class TFMap : public QObject {
```
Done

clyne replied:
The public methods aren't documented.

StasJ replied:
Added documentation/renamed functions to clarify what they do.

StasJ commented on apps/vaporgui/TFOpacityWidget.h:
```diff
@@ -0,0 +1,171 @@
+#include <vapor/VAssert.h>
+#include <glm/glm.hpp>
+
+class ControlPointList {
```
This is an internal component

StasJ commented on apps/vaporgui/TFOpacityWidget.h:
```diff
@@ -0,0 +1,171 @@
+    LineIterator EndLines()   { return LineIterator(this, SizeLines()); }
+};
+
+class TFOpacityMap : public TFMap {
```
I added documentation for TFMap. Here the documentation would only state "A TFMap for the opacity parameter" which doesn't add anything.

StasJ commented on apps/vaporgui/TFUtils.h:
```diff
@@ -0,0 +1,15 @@
+}
+
+namespace TFUtils {
+    void LoadColormap(VAPoR::MapperFunction *tf, const std::string &path);
```
No because we are modifying it by loading a new colormap into it.

StasJ commented on apps/vaporgui/TFUtils.h:
```diff
@@ -0,0 +1,15 @@
+
+namespace TFUtils {
+    void LoadColormap(VAPoR::MapperFunction *tf, const std::string &path);
+    void LoadColormap(VAPoR::ParamsMgr *paramsMgr, VAPoR::MapperFunction *tf);
```
Not without refactoring the params classes.

StasJ commented on apps/vaporgui/TFUtils.h:
```diff
@@ -0,0 +1,15 @@
+    class ParamsMgr;
+}
+
+namespace TFUtils {
```
The documentation would be "Utility functions for the TF". I don't think that adds anything other than redundancy

StasJ commented on apps/vaporgui/VSection.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <QToolButton>
+
+class VSection : public QTabWidget {
```
Done

StasJ commented on include/vapor/DataMgr.h:
```diff
@@ -460,6 +460,9 @@ std::vector <size_t> GetCRatios(string varname) const;
  ) ;
+    
+ 
+ int GetDefaultMetaInfoStride(std::string varname, int refinementLevel);
```
Done

StasJ commented on include/vapor/ParamsMgr.h:
```diff
@@ -468,6 +468,10 @@ class PARAMS_API ParamsMgr : public MyBase {
 	_ssave.EndGroup();
  };
+    
+    void IntermediateChange() {
```
Done

StasJ commented on include/vapor/ParamsMgr.h:
```diff
@@ -523,6 +527,10 @@ class PARAMS_API ParamsMgr : public MyBase {
 	_ssave.RegisterStateChangeCB(callback);
  }
+    
+    void RegisterIntermediateStateChangeCB(std::function<void()> callback) {
```
The documentation for the function right above, RegisterStateChangeCB, just says that it registers a callback that will be called when the state changes. That can be inferred from the function name although I think it could be improved by renaming it to RegisterIntermediateStateChangeCallbackFunction. You could argue that you should specify what an intermediate change is. However that has nothing to do with the job of this function. That should be described in the class description so I was going to do that but it turns out all of the documentation is split up amongst the functions so you have to read the documentation for each of the functions to figure out what the class does so I copied it from the IntermediateChange function.

StasJ commented on include/vapor/RenderParams.h:
```diff
@@ -366,6 +366,13 @@ class PARAMS_API RenderParams : public ParamsBase {
     bool secondaryColormapVariable
 );
+    
+    virtual bool HasIsoValues() const { return false; }
```
Done

StasJ commented on lib/vdc/DataMgr.cpp:
```diff
@@ -1693,6 +1693,26 @@ int DataMgr::GetDataRange(
+// Please I'm just trying to make things work.
+#define REQUIRED_SAMPLE_SIZE 1000000
+
+int DataMgr::GetDefaultMetaInfoStride(std::string varname, int refinementLevel)
```
Moved to DataMgrUtils. I don't know, I just moved this from the old TFWidget.

shaomeng commented:
Good widget! A few minor areas for improvement:

1. With the following menu configuration, the "diverging" option doesn't seem to change anything when it's checked/unchecked.
![Screenshot from 2019-10-07 15-38-33](https://user-images.githubusercontent.com/814885/66352212-6424cc00-e91c-11e9-9625-a468bdb81d45.png)

2. Sometimes it's hard to point to and click on the transfer function control points. It would improvement usability if the internal calculation considers a bigger area than the control points.

3. It seems I could input an arbitrary number in the box without having any impact on the rendering?
![Screenshot from 2019-10-07 16-05-42](https://user-images.githubusercontent.com/814885/66352409-afd77580-e91c-11e9-9011-5c58dc76d08d.png)

shaomeng commented on apps/vaporgui/Histo.h:
```diff
@@ -20,42 +20,70 @@
 //
```
The following code outputs a zero, while I suspect it should output a one.

```C++
#include <iostream>
#include "Histo.h"

int main()
{
    Histo h( 10, 0.0f, 1.0f, "haha", 1 );

    float v = 1.0f;
    h.addToBin( v );
    int idx = h.getBinIndexForValue( v );
    std::cout << h.getBinSize( idx ) << std::endl;
}
```

StasJ replied:
Not my code.

clyne replied:
@StasJ is Histo.h going to be deprecated?

StasJ replied:
I think it would be a good idea to replace it with something more robust

clyne replied:
Agreed. Let's leave it for future work

shaomeng commented:
Comment on the `Histo` class. 

This class performs standard histogram calculations, as well as interacts with VAPOR Data Manager to update itself with VAPOR variables.
I found it's cumbersome to have a single class perform these two distinct tasks, and the fusion of these tasks make it difficult to review the correctness of the calculations. (I found one behavior that I suspect it is wrong.)

I'd suggest separate these two functionalities, i.e., one histogram class that does only standard histogram calculations, and a separate class that connects a standard histogram class to various VAPOR parameters/Data Managers, etc. The 2nd class could play the role of what the current `Histo` is playing.

The good news is, we don't even need to implement a  standard histogram class and worry about its correctness: both GNU Scientific Library ([link](https://www.gnu.org/software/gsl/doc/html/histogram.html)) and boost ([link](https://www.boost.org/doc/libs/1_71_0/libs/histogram/doc/html/index.html)) have it for us.

StasJ commented on apps/vaporgui/GLColorbarWidget.cpp:
```diff
@@ -55,7 +55,7 @@ void GLColorbarWidget::setColormap(ColorMap *colormap)
   _colormap = colormap;
   vector<double> cps = colormap->GetControlPoints();
-  VAssert(cps.size()>0);
+  //VAssert(cps.size()>0);
```
This is the old TFWidget.

StasJ commented on apps/vaporgui/Histo.cpp:
```diff
@@ -17,140 +17,420 @@
-}
-#endif
 	
 Histo::~Histo(){
```
This is not my code. Also, the copy constructor will be removed once the old TFWidget is removed.

StasJ commented:
> Comment on the `Histo` class.
> 
> This class performs standard histogram calculations, as well as interacts with VAPOR Data Manager to update itself with VAPOR variables.
> I found it's cumbersome to have a single class perform these two distinct tasks, and the fusion of these tasks make it difficult to review the correctness of the calculations. (I found one behavior that I suspect it is wrong.)
> 
> I'd suggest separate these two functionalities, i.e., one histogram class that does only standard histogram calculations, and a separate class that connects a standard histogram class to various VAPOR parameters/Data Managers, etc. The 2nd class could play the role of what the current `Histo` is playing.
> 
> The good news is, we don't even need to implement a standard histogram class and worry about its correctness: both GNU Scientific Library ([link](https://www.gnu.org/software/gsl/doc/html/histogram.html)) and boost ([link](https://www.boost.org/doc/libs/1_71_0/libs/histogram/doc/html/index.html)) have it for us.

This code is not part of the PR. I agree that the Histo class doesn't make too much sense. We can open a separate feature request to fix these issues.

StasJ commented:
> Good widget! A few minor areas for improvement:
> 
> 1. With the following menu configuration, the "diverging" option doesn't seem to change anything when it's checked/unchecked.

Good catch. Fixed.

> 2. Sometimes it's hard to point to and click on the transfer function control points. It would improvement usability if the internal calculation considers a bigger area than the control points.

I implemented this but I discovered that it ended up not working too well. The original problem is caused by the fact that when looking at a mouse cursor, the white corner is what you're pointing to, not the black corner. Thus, when coming in from the bottom/right, the user may assume the mouse is on the object when it is not. Adding a buffer zone fixed this, however, now when going in from the left/top the user may assume they are outside of the object when they are not. So we could fix this by only having the buffer zone exist on the right/bottom. There is a problem: We then need to query whether the user is using a left handed cursor as the issue would be reversed. So I think we just expect the user to properly click on the object for now as this is nothing new.

> 3. It seems I could input an arbitrary number in the box without having any impact on the rendering?

That is how all of vapor works.
![s](https://user-images.githubusercontent.com/2772687/66611024-8451c680-eb7a-11e9-9106-0452d81d0c59.png)

StasJ commented on apps/vaporgui/QRangeSlider.cpp:
```diff
@@ -0,0 +1,237 @@
+#include <QBitmap>
+#include <QProxyStyle>
+
+#define QT_STOPS 1000000000
```
done

StasJ commented on apps/vaporgui/BarbSubtabs.cpp:
```diff
@@ -58,6 +59,7 @@ BarbAppearanceSubtab::BarbAppearanceSubtab(QWidget* parent) {
 	_TFWidget->Reinit(
 		(TFFlags)(CONSTANT_COLOR | COLORMAP_VAR_IS_IN_TF1)
 	);
+    verticalLayout->insertWidget(0, _tfe = new TFEditor);
```
The memory is managed by Qt

StasJ commented on apps/vaporgui/TFOpacityWidget.h:
```diff
@@ -0,0 +1,171 @@
+    void menuLoad();
+    void menuSave();
+};
+
```
It is reasonable but you could just CTRL-F "TFOpacityWidget". I can change it if you feel strongly about it.

clyne commented:
@StasJ please resolve the file conflicts and I'll accept this PR.

StasJ commented:
Done

On Thu, Oct 17, 2019 at 2:28 PM clyne <notifications@github.com> wrote:

> @StasJ <https://github.com/StasJ> please resolve the file conflicts and
> I'll accept this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/VAPOR/pull/1858?email_source=notifications&email_token=AAVE5T2MOLRFLKM4MBQXTTLQPDDFNA5CNFSM4I4ALHC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBRNL4A#issuecomment-543348208>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVE5T2J7GBHZTWP5OD4WNDQPDDFNANCNFSM4I4ALHCQ>
> .
>
